bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java

```diff
@@ -95,72 +100,153 @@ public Locale locale() {
-    /**
-     * @return a new DialogContext instance with the provided stt
-     */
-    public DialogContext withSTT(STTService stt) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided tts
-     */
-    public DialogContext withTTS(TTSService tts) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided voice
-     */
-    public DialogContext withVoice(Voice voice) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided list of interpreters
-     */
-    public DialogContext withHLIs(List<HumanLanguageInterpreter> hlis) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided locale
-     */
-    public DialogContext withLocale(Locale locale) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided listening item
-     */
-    public DialogContext withListeningItem(String listeningItem) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided ks
-     */
-    public DialogContext withKS(@Nullable KSService ks) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
-    }
-
-    /**
-     * @return a new DialogContext instance with the provided keyword
-     */
-    public DialogContext withKeyword(@Nullable String keyword) {
-        return new DialogContext(ks, keyword, stt, tts, voice, hlis, source, sink, locale, listeningItem);
+    public static class Builder {
+        private final Logger logger = LoggerFactory.getLogger(Builder.class);
+        // services
+        private @Nullable String sourceId = null;
+        private Supplier<@Nullable AudioSource> source = () -> null;
+        private @Nullable String sinkId = null;
+        private Supplier<@Nullable AudioSink> sink = () -> null;
+        private @Nullable String ksId = null;
+        private Supplier<@Nullable KSService> ks = () -> null;
+        private @Nullable String sttId = null;
+        private Supplier<@Nullable STTService> stt = () -> null;
+        private @Nullable String ttsId = null;
+        private Supplier<@Nullable TTSService> tts = () -> null;
+        private @Nullable Voice voice = null;
+        private List<String> hliIds = List.of();
+        private Supplier<List<HumanLanguageInterpreter>> hlis = List::of;
```

Comment:
I think the builder code could be simplified and will be easier to use if it does not use all these `Supplier`s.

Reply:
The reason to this is that the current implementation of the dialog processor allows to pass unregistered services to it, which is something I'm relaying on in the HABSpeaker addon (marketplace), with the suppliers this is posible.

As an explanation the addon manages the dialog creation disposal for each speaker (the web client) adding some unregistered services that are bounded to that speaker, the ability of non register those services (in my case a ks and a interpreter) allow me to not display to the user services that they should not use.

I can remove the ids if you want, those are there just to give context about the desired services, which can be different than the resolved ones for ks, tts, stt and hlis. I added the ids there also thinking in a future improvement of the voice manager.

So if it's ok for you I would prefer to keep the suppliers, but let me know if you think is not ok.

Reply:
I think I understand it incorrectly. If you meant remove the suppliers and just reference the services, yes I can change that without any impact and seems like a good idea.
I understand you were talking about just keeping the service ids.

Reply:
Yes like replacing `Supplier<@Nullable TTSService> tts` with `@Nullable TTSService tts` ... if that would work for you.

Reply:
The reason for me to did it like this was having some context about the desired services so the builder contain the information to rebuild the dialog later. 
But maybe I overthought it. 
Because right now you can start a dialog with a configured stt, it can fallback to a different one (because that one is not available at that moment) and the voice manager will have not clue that this dialog preference was using other stt service.
Anyway I will open a issue in the future reflecting all this thoughts sorry for the hesitation.

Reply:
The code looks much better now, thank you.
Seeing it in perspective I was mixing things, so really appreciated!